### PR TITLE
Fixes #12211 - Improve DHCP subnets parsing to get more informations …

### PR DESCRIPTION
**Pull request by bagasse**
…from DHCP server

Add ability to parse some subnet options and provides them to the foreman to enhance subnet imports.

This modify the dhcp smart-proxy API (adding options hash).

### CODE REVIEW

File: modules/dhcp/providers/server/isc.rb [hunk 1]
```diff
@@ -149,11 +149,24 @@ def loadSubnetData subnet
...
-
-        ret_val << Proxy::DHCP::Subnet.new(match[0], match[1])
+      @config.scan(/subnet\s+([\d\.]+)\s+netmask\s+([\d\.]+)\s*{([\w|;|\.|,|\-|\s]*)}/) do |match|
+        options = {}
+        if match[2]
+          match[2].split(';').each do |option|
+            case option
+              when /option\s+routers\s+[\d\.]+/
+                # get the first router match and use it for gateway
+                options[:gateway] = option.scan(/option\s+routers\s+([\d\.]+)/).first.first
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb]: why not keep the whole list?

File: modules/dhcp/providers/server/isc.rb [hunk 2]
```diff
@@ -149,11 +149,24 @@ def loadSubnetData subnet
...
+              when /option\s+domain\-name\-servers\s+[\d\.]+/
+                # get #1 and #2 domain name servers
+                options[:dns_primary], options[:dns_secondary] = option.scan(/option\s+domain\-name\-servers\s+([\d\.]+)\s*[,]?\s*([\d\.]*)/).first
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb]: similarly, why not keep the whole list? Also, this will leave `:dns_secondary` with an empty string if only one server has been specified.

**Comment by dmitri-d** [comment]
@bagasse: cheers for the PR!

I feel that in the data we return we should keep option names and format of their values close to the original ones (if not the same): `gateway` would become `:routers` and contain the whole list, `:dns_primary` and `:dns_secondary` would become `:domain_name_servers` and keep all of the servers, etc.

Another issue that I see is that dhcp config file may contain global options applicable to all subnets (and potentially overridden in individual subnet configs). I we should support this if we are going to return these additional options.

File: modules/dhcp/providers/server/isc.rb (hunk 2 above)
Comment (bagasse) [inline modules/dhcp/providers/server/isc.rb]: I reject all nil and blank values just before return the array

Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb]: nods, see it now.

**Comment by bagasse** [comment]
@witlessbird

I changed the options names to map names used on the foreman side. I will update the PR to keep names like in the dhcp server config, and keep all values too.
For the global options, i didn't take a look at it because you can have global options, shared networks blocks and group block (with some nested ones) that overrides some of the options. I wanted to keep this simple at this time and just take local subnet options. Parsing dhcpd config with regex with this kind of russian dolls config blocks was out of the scope of this PR (and testing all cases is not trivial), but if it's a requirement for you, i will take a look at it when i will have more time.

**Comment by dmitri-d** [comment]
@bagasse: let's wait with the renaming until we hear something from the foreman side (although I think it makes sense to rename attribute names there and keep them close to original here).

Yeah, I agree that parsing global options makes things more complicated (and possibly warrants a real parser, not regexes). I'm ok with postponing this bit, as long as we can get the functionality in not too distant future.

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
@bagasse One of the tests you modified (get_root) is failing. see [Jenkins](http://ci.theforeman.org/job/test_proxy_develop_pull_request/888/)  or test locally with `rake test`. Rubocop is also throwing a few warnings, please fix them & update the PR.

**Comment by bagasse** [comment]
@dLobatog: I've fixed rubocop warnings, sorry i forgot to run rubocop before on this PR. For the get_root  of dhcp_api_test, all test are green in my dev env (CentOS 7, ruby 1.9.3). I seen in jenkins that this test only fail with ruby 1.8.7. I don't know why, with this ruby version, networks presents in the list are not in the same order than in others ruby versions. I will take a look at this more deeply tomorrow.

**Comment by dmitri-d** [comment]
[test]

File: modules/dhcp/providers/server/isc.rb [hunk 3]
```diff
@@ -288,8 +298,7 @@ def read_config file, ignore_includes=false
       logger.debug "Reading config file #{file}"
       config = []
       File.readlines(file).each do |line|
-
-        line.strip! # remove left and right whitespace
+        line = line.split('#').first.strip # remove comments, left and right whitespace
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb:305]: Why this change? The next line used to skip over comments (and isn't needed with this change).

File: test/dhcp/dhcp_api_test.rb [hunk 4]
```diff
@@ -37,7 +37,26 @@ def test_api_01_get_root
...
+        }
+      },{
+        "network"=>"192.168.124.0",
+        "netmask"=>"255.255.255.0",
+        "options"=>{
+          "routers"=>["192.168.124.1", "192.168.124.2"],
+          "domain_name_servers"=>["192.168.123.1", "192.168.122.250"]
+        }
+    }]
     assert_equal expected, data
```
Comment (dmitri-d) [inline test/dhcp/dhcp_api_test.rb]: Test failure here is due to the order of entries in `epxected` array being different from the order of entries in `data`. I would suggest using sets for equality comparison (we don't care about the order). You can find an example in other tests in this suite.

File: modules/dhcp/providers/server/isc.rb (hunk 3 above)
Comment (bagasse) [inline modules/dhcp/providers/server/isc.rb:305]: In my understanding, the next line remove lines that starts with comment, this remove comments at the end of lines. But if you want i can remove this.

**Comment by bagasse** [comment]
@witlessbird: changed, arrays to sets, thanks for pointing this out to me.

**Comment by dLobatog** [comment]
[test]

File: modules/dhcp/subnet.rb [hunk 5]
```diff
@@ -17,11 +17,19 @@ class Subnet
     include Proxy::Log
     include Proxy::Validations
 
-    def initialize network, netmask
-      @network   = validate_ip network
-      @netmask   = validate_ip netmask
-      @options   = {}
-      @timestamp = Time.now
+    def initialize network, netmask, options = {}
+      @network       = validate_ip network
```
Comment (dmitri-d) [inline modules/dhcp/subnet.rb]: nitpick: you can get rid of assignment formatting -- simple spaces around '=' would suffice.

File: modules/dhcp/providers/server/isc.rb (hunk 3 above)
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb:305]: once you call `String#split` all instances of the symbol you used to split the string will be removed,  which makes https://github.com/bagasse/smart-proxy/blob/Fixes_12211_-_Improve_DHCP_subnets_parsing_to_get_more_informations_from_DHCP_server/modules/dhcp/providers/server/isc.rb#L302 redundant. 

I'm ok with this change if you remove code on line 302.

File: test/dhcp/server_isc_test.rb [hunk 6]
```diff
@@ -82,8 +82,24 @@ def test_poap_quirks
   def test_loadSubnets_loads_managed_subnets
     subnets = @dhcp.loadSubnets
```
Comment (dmitri-d) [inline test/dhcp/server_isc_test.rb]: I know that this was here before, but it wasn't a very good test: `loadSubnets` returns subnets that were processed, not the data that was stored.

Another issue is that this tests too many things; which will make it harder to pinpoint an issue should code change. Could you break this up to test specific cases, for example handling of `domain_name_server` option when it's not present, when only one server is present, when both are present, etc. 

It is probably easier to work directly with subnet_service (see `@subnet_service` in setup -- you can query it for relevant subnets, as opposed to parsing the hash returned by `@dhcp.subnets`.

File: modules/dhcp/providers/server/isc.rb [hunk 7]
```diff
@@ -409,4 +418,11 @@ def poap_options_statements(options)
       statements
     end
   end
+
+  private
+
+  def get_ip_list_from_config_line(option_line)
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb:426]: can you add a test for this method (you'll need to drop `private` for this)?

File: test/dhcp/subnet_test.rb [hunk 8]
```diff
@@ -24,6 +23,33 @@ def test_should_not_save_invalid_network_addresses
     assert_raise Proxy::Validations::Error do
       Proxy::DHCP::Subnet.new("1..1.1", @netmask)
     end
+    assert_raise Proxy::Validations::Error do
```
Comment (dmitri-d) [inline test/dhcp/subnet_test.rb:29]: Can you change this to test per option?

**Comment by dmitri-d** [comment]
@bagasse: Looks pretty good, only needs some changes in tests around added functionality.

**Comment by domcleal** [comment]
Thanks for your PRs and work on this feature @bagasse, and welcome to our GitHub organisation.  If you can set your membership to "public" on [this page](https://github.com/orgs/theforeman/people) then Jenkins will automatically test any updates you make.

File: modules/dhcp/providers/server/isc.rb (hunk 3 above)
Comment (bagasse) [inline modules/dhcp/providers/server/isc.rb:305]: line 302 removed.

File: modules/dhcp/subnet.rb (hunk 5 above)
Comment (bagasse) [inline modules/dhcp/subnet.rb]: done.

File: test/dhcp/server_isc_test.rb (hunk 6 above)
Comment (bagasse) [inline test/dhcp/server_isc_test.rb]: In my understanding, `loadSubnets` implementation of `Proxy::DHCP::Server::ISC` load subnets only from config files and `Proxy::DHCP::SubnetService` is only here to store loaded data, so i don't see the problem to use `loadSubnets` here. For the other point, i've separated tests in several methods.

File: modules/dhcp/providers/server/isc.rb (hunk 7 above)
Comment (bagasse) [inline modules/dhcp/providers/server/isc.rb:426]: done some basic test for this method

File: test/dhcp/subnet_test.rb (hunk 8 above)
Comment (bagasse) [inline test/dhcp/subnet_test.rb:29]: moved per options test into other methods

**Comment by domcleal** [comment]
[test] one last time :)

**Comment by bagasse** [comment]
@domcleal Glad to join and thank you for the invitation :)

File: modules/dhcp/providers/server/isc.rb [hunk 9]
```diff
@@ -149,11 +148,22 @@ def loadSubnetData subnet
...
     def parse_config_for_subnets
       ret_val = []
-
-      @config.scan(/subnet\s+([\d\.]+)\s+netmask\s+([\d\.]+)/) do |match|
-        next unless managed_subnet? "#{match[0]}/#{match[1]}"
-
-        ret_val << Proxy::DHCP::Subnet.new(match[0], match[1])
+      @config.scan(/subnet\s+([\d\.]+)\s+netmask\s+([\d\.]+)\s*{([\w|;|\.|,|\-|\s]*)}/) do |match|
+        options = {}
+        if match[2]
```
Comment (dLobatog) [inline modules/dhcp/providers/server/isc.rb]: Mind to extract this logic to a method or something similar? Maybe write a comment / save `match[0] match[1] match[2]` to named variables? It's a bit difficult to understand without being well acquainted with the format of the conf file.

File: modules/dhcp/providers/server/isc.rb [hunk 10]
```diff
@@ -149,11 +148,22 @@ def loadSubnetData subnet
...
+          match[2].split(';').each do |option|
+            case option
+              when /option\s+routers\s+[\d\.]+/
+                options[:routers] = get_ip_list_from_config_line(option)
+              when /option\s+domain\-name\-servers\s+[\d\.]+/
+                options[:domain_name_servers] = get_ip_list_from_config_line(option)
+              when /range\s+[\d\.]+\s+[\d\.]+/
+                # get IP addr range used for this subnet
+                options[:range] = option.scan(/\s*((?:\d{1,3}\.){3}\d{1,3})\s*((?:\d{1,3}\.){3}\d{1,3})\s*/).first
```
Comment (dLobatog) [inline modules/dhcp/providers/server/isc.rb]: Perhaps you want to extract this regex to a method `ip_address_range`? Maybe move the regex to a constant?
 You use almost all of it below as well on `get_ip_list_from_config_line`, and it's hard to know what it is without a comment, maybe that would help making this easier to read.

**Comment by dLobatog** [comment]
@bagasse I only have 2 comments about the regexes & maintainability, but overall great job :+1:. IMO if you could fix that that should be it. In any case @domcleal and @witlessbird feel free to merge if you think it's fine already, you are the ones that work on this code the most.

**Comment by bagasse** [comment]
@dLobatog you are right, it's not really readable, i've extracted range regexp too and used named vars.

File: test/dhcp/server_isc_test.rb (hunk 6 above)
Comment (dmitri-d) [inline test/dhcp/server_isc_test.rb]: Nods,that's correct. However, currently these tests verify the behaviour of `parse_config_for_subnets` method: so either calls to `loadSubnets`should be replaced with calls to `parse_config_for_subnets`, or the assertions should be made against `Proxy::DHCP::SubnetService` (the main side-effect of `loadSubnets` is initialization of `SubnetService`).

File: modules/dhcp/providers/server/isc.rb [hunk 11]
```diff
@@ -409,4 +421,14 @@ def poap_options_statements(options)
       statements
     end
   end
+
+  # Get all IPv4 addresses provided by the ISC DHCP config line following the pattern "my-config-option-or-directive IPv4_ADDR[, IPv4_ADDR] [...];" and return an array.
+  def get_ip_list_from_config_line(option_line)
```
Comment (dmitri-d) [inline modules/dhcp/providers/server/isc.rb:426]: Could you add a test for this method?

Comment (bagasse) [inline modules/dhcp/providers/server/isc.rb:426]: It was already there, but i've added a test for `get_range_from_config_line` that i've extracted too.  I changed `loadSubnets` calls to `parse_config_for_subnets` as suggested.

**Comment by dmitri-d** [comment]
[test]

**Comment by dmitri-d** [comment]
Waiting for tests to pass before merging.

**Comment by dmitri-d** [comment]
merged in at e2a21ec. Thanks, @bagasse!

**Comment by bagasse** [comment]
@witlessbird @dLobatog, thank you for pointers and re (re re re)-reviews :)
